src/main/java/org/candlepin/subscriptions/tally/roller/BaseSnapshotRoller.java

```diff
@@ -0,0 +1,115 @@
+
+import org.candlepin.subscriptions.db.TallySnapshotRepository;
+import org.candlepin.subscriptions.db.model.AccountMaxValues;
+import org.candlepin.subscriptions.db.model.TallyGranularity;
+import org.candlepin.subscriptions.db.model.TallySnapshot;
+import org.candlepin.subscriptions.util.ApplicationClock;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.OffsetDateTime;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+/**
+ * Base class for all usage snapshot rollers. A snapshot roller is responsible compressing
+ * finer granularity snapshots into more compressed snapshots. For example, rolling daily
+ * snapshots into weekly snapshots or rolling weekly snapshots into monthly snapshots.
+ */
+public abstract class BaseSnapshotRoller {
+
+    private static final Logger log = LoggerFactory.getLogger(BaseSnapshotRoller.class);
+
+    protected  TallySnapshotRepository tallyRepo;
+    protected ApplicationClock clock;
+    protected String product;
+
+    public BaseSnapshotRoller(String product, TallySnapshotRepository tallyRepo,
+        ApplicationClock clock) {
+        this.tallyRepo = tallyRepo;
+        this.clock = clock;
+        this.product = product;
+    }
+
+    /**
+     * Roll the snapshots for the given account.
+     *
+     * @param accounts the accounts of the snapshots to roll.
+     */
+    public abstract void rollSnapshots(List<String> accounts);
+
+    protected TallySnapshot createSnapshotFromMaxValues(AccountMaxValues maxValues,
+        TallyGranularity granularity) {
+        TallySnapshot snapshot = new TallySnapshot();
+        snapshot.setSnapshotDate(getSnapshotDate(granularity));
+        snapshot.setInstanceCount(maxValues.getMaxInstances());
+        snapshot.setCores(maxValues.getMaxCores());
+        snapshot.setProductId(this.product);
+        snapshot.setGranularity(granularity);
+        snapshot.setOwnerId(maxValues.getOwnerId());
+        snapshot.setAccountNumber(maxValues.getAccountNumber());
+        return snapshot;
+    }
+
+    protected TallySnapshot updateWithMax(TallySnapshot snapshot, AccountMaxValues maxValues) {
+        snapshot.setInstanceCount(maxValues.getMaxInstances());
+        snapshot.setCores(maxValues.getMaxCores());
+        snapshot.setOwnerId(maxValues.getOwnerId());
+        snapshot.setAccountNumber(maxValues.getAccountNumber());
+        snapshot.setSnapshotDate(getSnapshotDate(snapshot.getGranularity()));
+        return snapshot;
+    }
+
+    protected void updateSnapshots(Map<String, TallySnapshot> existingSnapsByAccount,
+        Map<String, AccountMaxValues> maxValuesByAccount, TallyGranularity targetGranularity) {
+        List<TallySnapshot> snapshots = new LinkedList<>();
+        for (Entry<String, AccountMaxValues> next : maxValuesByAccount.entrySet()) {
+            TallySnapshot snap = existingSnapsByAccount.containsKey(next.getKey()) ?
```

Comment:
Can we break this down into an `if` block?  This is pretty dense and separating it out a little might make it easier to parse, mentally.

Reply:
DONE